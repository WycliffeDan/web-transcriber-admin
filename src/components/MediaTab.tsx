import React, { useState, useEffect } from 'react';
import { useGlobal } from 'reactn';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import * as actions from '../actions';
import { IState, IUploadFile, MediaFile, Passage, PassageSection,
  Section, IMediaTabStrings } from '../model';
import localStrings from '../selector/localize';
import { withData } from 'react-orbitjs';
import { KeyMap, QueryBuilder } from '@orbit/data';
import { withStyles, WithStyles, Theme } from '@material-ui/core/styles';
import { Button, Menu, MenuItem, IconButton } from '@material-ui/core'
import DropDownIcon from '@material-ui/icons/ArrowDropDown';
import AddIcon from '@material-ui/icons/Add';
import FilterIcon from '@material-ui/icons/FilterList';
import SelectAllIcon from '@material-ui/icons/SelectAll';
import MediaUpload from './MediaUpload';
import SnackBar from './SnackBar';
import Confirm from './AlertDialog';
import ShapingTable from './ShapingTable';
import related from '../utils/related';
import Auth from '../auth/Auth';
import moment from 'moment';
import 'moment/locale/fr';

const styles = (theme: Theme) => ({
  container: {
      display: 'flex',
      marginLeft: theme.spacing.unit * 4,
      marginRight: theme.spacing.unit * 4,
      marginBottom: theme.spacing.unit * 4,
  },
  paper: {
  },
  actions: theme.mixins.gutters({
    paddingBottom: 16,
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'right'
  }),
  button: {
    margin: theme.spacing.unit
  },
  icon: {
    marginLeft: theme.spacing.unit
  },
});

interface IRow {
  fileName: string;
  section: string;
  reference: string;
  duration: string;
  size: number;
  version: string;
  date: string;
};

const getSection = (section: Section[]) => {
  const sectionId = (section.length > 0 && section[0].attributes.sequencenum)? section[0].attributes.sequencenum.toString(): '';
  const sectionName = section.length > 0? section[0].attributes.name: '';
  return sectionId + " " + sectionName;
}

const getReference = (passage: Passage[]) => {
  const book = passage.length > 0? passage[0].attributes.book: '';
  const reference = passage.length > 0? passage[0].attributes.reference: '';
  return book + " " + reference;
}

interface ILatest {
  [name: string]: number;
}

const getMedia = (plan: string, mediaFiles: Array<MediaFile>, passages: Array<Passage>, passageSections: Array<PassageSection>, sections: Array<Section>) => {
  const latest: ILatest = {};
  mediaFiles.forEach(f => {
    const name = f.attributes.originalFile;
    latest[name] = latest[name]? Math.max(latest[name], f.attributes.versionNumber):
      f.attributes.versionNumber;
  })
  const media = mediaFiles.filter(f => related(f, 'plan') === plan &&
    latest[f.attributes.originalFile] === f.attributes.versionNumber);
  const rowData = media.map(f => {
    const passageId = related(f, 'passage');
    const passage = passages.filter(p => p.id === passageId);
    const passageSection = passageSections.filter(ps => related(ps, 'passage') === passageId);
    const sectionId = passageSection.length > 0? related(passageSection[0], 'section'): '';
    const section = sections.filter(s => s.id === sectionId);
    const updated = f.attributes.dateUpdated && moment(f.attributes.dateUpdated + 'Z');
    const date = updated? updated.format('YYYY-MM-DD'): ''
    const displayDate = updated? updated.locale(navigator.language.split('-')[0]).format('L'):'';
    const displayTime = updated? updated.locale(navigator.language.split('-')[0]).format('LT'):'';
    const today = moment().format('YYYY-MM-DD')
    return {
      fileName: f.attributes.originalFile,
      section: getSection(section),
      reference: getReference(passage),
      duration: f.attributes.duration? f.attributes.duration.toString(): '',
      size: f.attributes.filesize,
      version: f.attributes.versionNumber? f.attributes.versionNumber.toString(): '',
      date: date === today? displayTime: displayDate,
    } as IRow
  });
  return(rowData as Array<IRow>);
};

interface IStateProps {
  t: IMediaTabStrings;
  uploadList: Array<IUploadFile>;
  loaded: boolean;
  currentlyLoading: number;
};

interface IDispatchProps {
  uploadFiles: typeof actions.uploadFiles;
  nextUpload: typeof actions.nextUpload;
}

interface IRecordProps {
  mediaFiles: Array<MediaFile>;
  passages: Array<Passage>;
  passageSections: Array<PassageSection>;
  sections: Array<Section>;
}

interface IProps extends IStateProps, IDispatchProps, IRecordProps, WithStyles<typeof styles>{
  action?: (what: string, where: number[]) => boolean;
  auth: Auth;
};
  
export function MediaTab(props: IProps) {
  const { classes, t, uploadList, loaded, currentlyLoading,
    action, uploadFiles, nextUpload,
    mediaFiles, passages, passageSections, sections, auth } = props;
  const [plan] = useGlobal('plan');
  const [keyMap] = useGlobal('keyMap');
  const [message, setMessage] = useState(<></>);
  const [data, setData] = useState(Array<IRow>());
    // [
    //   {fileName: 'GEN-001-001025.mp3', sectionId: '1', sectionName: 'Creation Story', book: 'Genesis', reference: '1:1-25a', duration: '30 seconds', size: 250, version: '1' },
    //   {fileName: 'GEN-001-002631.mp3', sectionId: '', sectionName: '', book: '', reference: '', duration: '45 seconds', size: 445, version: '1' },
    // ]);
  const [actionMenuItem, setActionMenuItem] = useState(null);
  const [check, setCheck] = useState(Array<number>());
  const [confirmAction, setConfirmAction] = useState('');
  const columnDefs=[
    { name: 'fileName', title: t.fileName },
    { name: 'section', title: t.section },
    { name: 'reference', title: t.reference },
    { name: 'duration', title: t.duration },
    { name: 'size', title: t.size },
    { name: 'version', title: t.version },
    { name: 'date', title: t.date }
  ];
  const columnWidths=[
    { columnName: "fileName", width: 150 },
    { columnName: "section", width: 150 },
    { columnName: "reference", width: 150 },
    { columnName: "duration", width: 100 },
    { columnName: "size", width: 100 },
    { columnName: "version", width: 100 },
    { columnName: "date", width: 100 }
  ];
  const numCompare = (a:number, b:number) => {return a-b};
  const dateCompare = (a:string, b:string) => {
    const aDate = moment(a).isValid()? moment(a): moment(a, 'LT');
    const bDate = moment(b).isValid()? moment(b): moment(b, 'LT');
    const aIso = aDate.toISOString();
    const bIso = bDate.toISOString();
    return aIso > bIso? 1:
      aIso < bIso? -1: 0
  }
  const columnSorting=[
    { columnName: 'duration', compare: numCompare },
    { columnName: 'size',     compare: numCompare },
    { columnName: 'version',  compare: numCompare },
    { columnName: 'date',     compare: dateCompare },
  ];
  const numCols= ['duration','size','version'];
  const [filter, setFilter] = useState(false);
  const [uploadVisible, setUploadVisible] = useState(false);


  const handleMessageReset = () => { setMessage(<></>) }
  const handleUpload = () => { setUploadVisible(true) }
  const uploadMedia = (files: Array<IUploadFile>) => {
    uploadFiles(files);
    setUploadVisible(false);
  }
  const uploadCancel = () => { setUploadVisible(false) }
  const handleMenu = (e:any) => setActionMenuItem(e.currentTarget);
  const handleConfirmAction = (what: string) => (e:any) => {
    setActionMenuItem(null)
    if (check.length === 0) {
      setMessage(<span>Please select row(s) to {what}.</span>)
    } else if (!/Close/i.test(what)) {
      setConfirmAction(what);
    }
  };
  const handleActionConfirmed = () => {
    if (action != null) {
      if (action(confirmAction, check)) {
        setCheck(Array<number>());
      }
    }
    setConfirmAction('');
  };
  const handleActionRefused = () => {
    setConfirmAction('');
  };
  const handleCheck = (checks: Array<number>) => {
    setCheck(checks);
  };
  const handleFilter = () => setFilter(!filter);

  useEffect(() => {
    setData(
      getMedia(plan as string, mediaFiles, passages, passageSections, sections)
    );
  }, [plan, mediaFiles, passages, passageSections, sections])

  useEffect(() => {
    if (loaded || currentlyLoading < 0) {
      if (uploadList.length > 0 && currentlyLoading + 1 < uploadList.length) {
        const planId = parseInt((keyMap as KeyMap).idToKey('plan', 'remoteId', (plan as string)));
        const mediaFile = {
            PlanId: planId,
          } as any;
        nextUpload(JSON.stringify(mediaFile), uploadList, currentlyLoading + 1, auth)
      }
    }
  /* eslint-disable-next-line react-hooks/exhaustive-deps */
  }, [uploadList, loaded, currentlyLoading, plan, auth])

  return (
    <div className={classes.container}>
      <div className={classes.paper}>
        <div className={classes.actions}>
        <Button
            key="action"
            aria-owns={actionMenuItem !== ''? 'action-menu': undefined}
            aria-label={t.action}
            variant="outlined"
            color="primary"
            className={classes.button}
            onClick={handleMenu}
          >
            {t.action}
            <DropDownIcon className={classes.icon} />
          </Button>
          <Menu
            id='action-menu'
            anchorEl={actionMenuItem}
            open={Boolean(actionMenuItem)}
            onClose={handleConfirmAction('Close')}
          >
            <MenuItem onClick={handleConfirmAction('Delete')}>{t.delete}</MenuItem>
            <MenuItem onClick={handleConfirmAction('Download')}>{t.download}</MenuItem>
            <MenuItem onClick={handleConfirmAction('Change Version')}>{t.changeVersion}</MenuItem>
            <MenuItem onClick={handleConfirmAction('Assign Passage')}>{t.assignPassage}</MenuItem>
          </Menu>
          <Button
            key="upload"
            aria-label={t.uploadMedia}
            variant="outlined"
            color="primary"
            className={classes.button}
            onClick={handleUpload}
          >
            {t.uploadMedia}
            <AddIcon className={classes.icon} />
          </Button>
          <IconButton onClick={handleFilter} title={"Show/Hide filter rows"}>
            {filter? <SelectAllIcon/>: <FilterIcon/>}
          </IconButton>
        </div>
        <ShapingTable
          columns={columnDefs}
          columnWidths={columnWidths}
          columnSorting={columnSorting}
          numCols={numCols}
          rows={data}
          select={handleCheck}
          shaping={filter}
        />
      </div>
      <MediaUpload
        visible={uploadVisible}
        uploadMethod={uploadMedia}
        cancelMethod={uploadCancel}
      />
      {confirmAction !== ''
      ? <Confirm
          text={confirmAction + ' ' + check.length + ' Item(s). Are you sure?'}
          yesResponse={handleActionConfirmed}
          noResponse={handleActionRefused}
        />
      : <></>}
      <SnackBar {...props} message={message} reset={handleMessageReset} />
    </div>
  );
}

const mapStateToProps = (state: IState): IStateProps => ({
  t: localStrings(state, {layout: "mediaTab"}),
  uploadList: state.upload.files,
  currentlyLoading: state.upload.current,
  loaded: state.upload.loaded,
});

const mapDispatchToProps = (dispatch: any): IDispatchProps => ({
  ...bindActionCreators({
    uploadFiles: actions.uploadFiles,
    nextUpload: actions.nextUpload,
  }, dispatch),
});

const mapRecordsToProps = {
  mediaFiles: (q: QueryBuilder) => q.findRecords('mediafile'),
  passages: (q: QueryBuilder) => q.findRecords('passage'),
  passageSections: (q: QueryBuilder) => q.findRecords('passagesection'),
  sections: (q: QueryBuilder) => q.findRecords('section'),
}

export default withStyles(styles, { withTheme: true })(
    withData(mapRecordsToProps)(
        connect(mapStateToProps, mapDispatchToProps)(MediaTab) as any
        ) as any
    ) as any;
      