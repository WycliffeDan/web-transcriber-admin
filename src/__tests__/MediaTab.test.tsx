import React from 'react';
import { Router } from 'react-router-dom';
import { setGlobal } from 'reactn';
import { DataProvider } from 'react-orbitjs';
import { Provider } from 'react-redux';
import store from '../store';
import { createMuiTheme } from '@material-ui/core';
import Store from '@orbit/store';
import { schema, keyMap } from '../schema';
import history from '../history';
import { render, cleanup, waitForElement } from '@testing-library/react';
import 'jest-dom/extend-expect';
import MediaTab from '../components/MediaTab';

const theme = createMuiTheme({});

const dataStore = new Store({ schema, keyMap });

const globals = {
  organization: null,
  project: null,
  plan: null,
  user: null,
  initials: null,
  lang: 'en',
  dataStore: dataStore,
  schema: schema,
  keyMap: keyMap,
};
setGlobal(globals);

const tree = (props?: any) => (
  <DataProvider dataStore={dataStore}>
    <Provider store={store}>
      <Router history={history}>
        <MediaTab {...props} />
      </Router>
    </Provider>
  </DataProvider>
);

const addMediaFile = async () => {
  const mediaFile = {
    type: 'mediafile',
    attributes: {
      originalFile: 'PAT-LUK-001-001004v01.mp3',
    },
  } as any;
  schema.initializeRecord(mediaFile);
  await dataStore.update(t => t.addRecord(mediaFile));
  return mediaFile.id as string;
};

const addPassageAndSection = async (mediaFileId: string) => {
  const passage = {
    type: 'passage',
    attributes: {
      sequencenum: 1,
      book: 'Luke',
      reference: '1:1-4',
      position: 0,
      state: 'initial',
      hold: false,
      title: 'Introduction',
    },
  } as any;
  schema.initializeRecord(passage);
  await dataStore.update(t => t.addRecord(passage));
  await dataStore.update(t =>
    t.replaceRelatedRecord({ type: 'mediafile', id: mediaFileId }, 'passage', {
      type: 'passage',
      id: passage.id,
    })
  );
  const section = {
    type: 'section',
    attributes: {
      name: 'Jesus is coming',
    },
  } as any;
  schema.initializeRecord(section);
  await dataStore.update(t => t.addRecord(section));
  const passageSection = {
    type: 'passagesection',
  } as any;
  schema.initializeRecord(passageSection);
  await dataStore.update(t => t.addRecord(passageSection));
  await dataStore.update(t =>
    t.replaceRelatedRecord(
      { type: 'passagesection', id: passageSection.id },
      'passage',
      { type: 'passage', id: passage.id }
    )
  );
  await dataStore.update(t =>
    t.replaceRelatedRecord(
      { type: 'passagesection', id: passageSection.id },
      'section',
      { type: 'section', id: section.id }
    )
  );
};

afterEach(cleanup);

test('can render MediaTab snapshot', async () => {
  const { getByText, container } = render(tree());
  const TestMediaTab = await waitForElement(() => getByText(/^File Name$/i));
  expect(container.firstChild).toMatchSnapshot();
  expect(getByText(/No Data/i)).toHaveTextContent('No data');
});

test('media files listed even if no passage', async () => {
  await addMediaFile();

  const { getByText } = render(tree());
  const TestMediaTab = await waitForElement(() => getByText(/^File Name$/i));
  expect(getByText(/PAT-LUK-001-001004v01.mp3/i)).toHaveTextContent(
    'PAT-LUK-001-001004v01.mp3'
  );
});

test('media files passage and section data listed if present', async () => {
  const mediaFileId = await addMediaFile();
  await addPassageAndSection(mediaFileId);

  const { getByText, getAllByText } = render(tree());
  const TestMediaTab = await waitForElement(() =>
    getAllByText(/^PAT-LUK-001-001004v01.mp3$/i)
  );
  expect(getByText(/Jesus is coming/i)).toHaveTextContent('Jesus is coming');
});
