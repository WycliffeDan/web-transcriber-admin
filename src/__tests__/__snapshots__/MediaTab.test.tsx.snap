// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can render MediaTab snapshot 1`] = `
<div
  class="WithData-Connect-MediaTab---container-1"
>
  <div
    class="WithData-Connect-MediaTab---paper-2"
  >
    <div
      class="WithData-Connect-MediaTab---actions-3"
    >
      <button
        aria-label="Action"
        aria-owns="action-menu"
        class="MuiButtonBase-root-32 MuiButton-root-6 MuiButton-outlined-14 MuiButton-outlinedPrimary-15 WithData-Connect-MediaTab---button-4"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label-7"
        >
          Action
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-35 WithData-Connect-MediaTab---icon-5"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root-156"
        />
      </button>
      <button
        aria-label="Upload Media"
        class="MuiButtonBase-root-32 MuiButton-root-6 MuiButton-outlined-14 MuiButton-outlinedPrimary-15 WithData-Connect-MediaTab---button-4"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label-7"
        >
          Upload Media
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-35 WithData-Connect-MediaTab---icon-5"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root-156"
        />
      </button>
      <button
        class="MuiButtonBase-root-32 MuiIconButton-root-48"
        tabindex="0"
        title="Show/Hide filter rows"
        type="button"
      >
        <span
          class="MuiIconButton-label-53"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-35"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 18h4v-2h-4v2zM3 6v2h18V6H3zm3 7h12v-2H6v2z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root-156"
        />
      </button>
    </div>
    <div
      class="RootBase-root-54"
    >
      <div
        class="TableContainer-root-55"
      >
        <table
          class="MuiTable-root-60 Table-table-56"
          style="min-width: 908px;"
        >
          <colgroup>
            <col
              style="width: 58px;"
            />
            <col
              style="width: 150px;"
            />
            <col
              style="width: 150px;"
            />
            <col
              style="width: 150px;"
            />
            <col
              style="width: 100px;"
            />
            <col
              style="width: 100px;"
            />
            <col
              style="width: 100px;"
            />
            <col
              style="width: 100px;"
            />
            <col />
          </colgroup>
          <thead
            class="MuiTableHead-root-61"
          >
            <tr
              class="MuiTableRow-root-62 MuiTableRow-head-65"
            >
              <th
                class="MuiTableCell-root-71 MuiTableCell-head-72 MuiTableCell-paddingCheckbox-77 TableSelectAllCell-cell-67"
                colspan="1"
                scope="col"
              >
                <span
                  class="MuiButtonBase-root-32 MuiButtonBase-disabled-33 MuiIconButton-root-48 MuiIconButton-disabled-52 MuiPrivateSwitchBase-root-89 MuiCheckbox-root-83 MuiCheckbox-colorSecondary-88 MuiPrivateSwitchBase-disabled-91 MuiCheckbox-disabled-85 TableSelectAllCell-checkbox-68"
                  tabindex="-1"
                >
                  <span
                    class="MuiIconButton-label-53"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root-35"
                      focusable="false"
                      role="presentation"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                      />
                    </svg>
                    <input
                      class="MuiPrivateSwitchBase-input-92"
                      data-indeterminate="false"
                      disabled=""
                      type="checkbox"
                      value=""
                    />
                  </span>
                </span>
              </th>
              <th
                class="MuiTableCell-root-71 MuiTableCell-head-72 TableHeaderCell-cell-94 TableHeaderCell-cellNoUserSelect-98 TableHeaderCell-cellDraggable-99 TableHeaderCell-cellNoWrap-104"
                colspan="1"
                scope="col"
              >
                <div
                  class="TableHeaderCell-container-103"
                >
                  <div
                    class="Content-content-105"
                  >
                    <div
                      class="SortLabel-root-108"
                    >
                      <span
                        class="MuiButtonBase-root-32 MuiTableSortLabel-root-121 SortLabel-sortLabelRoot-110"
                        role="button"
                        tabindex="0"
                        title="Sort"
                      >
                        <span
                          class="Title-title-126"
                        >
                          File Name
                        </span>
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root-35 MuiTableSortLabel-icon-123 MuiTableSortLabel-iconDirectionDesc-124"
                          focusable="false"
                          role="presentation"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                          />
                        </svg>
                      </span>
                    </div>
                  </div>
                </div>
                <div
                  class="ResizingControl-resizeHandle-127 TableHeaderCell-resizeHandle-95"
                >
                  <div
                    class="TableHeaderCell-resizeHandleLine-96 ResizingControl-resizeHandleLine-128 ResizingControl-resizeHandleFirstLine-129"
                  />
                  <div
                    class="TableHeaderCell-resizeHandleLine-96 ResizingControl-resizeHandleLine-128 ResizingControl-resizeHandleSecondLine-130"
                  />
                </div>
              </th>
              <th
                class="MuiTableCell-root-71 MuiTableCell-head-72 TableHeaderCell-cell-94 TableHeaderCell-cellNoUserSelect-98 TableHeaderCell-cellDraggable-99 TableHeaderCell-cellNoWrap-104"
                colspan="1"
                scope="col"
              >
                <div
                  class="TableHeaderCell-container-103"
                >
                  <div
                    class="Content-content-105"
                  >
                    <div
                      class="SortLabel-root-108"
                    >
                      <span
                        class="MuiButtonBase-root-32 MuiTableSortLabel-root-121 SortLabel-sortLabelRoot-110"
                        role="button"
                        tabindex="0"
                        title="Sort"
                      >
                        <span
                          class="Title-title-126"
                        >
                          Section
                        </span>
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root-35 MuiTableSortLabel-icon-123 MuiTableSortLabel-iconDirectionDesc-124"
                          focusable="false"
                          role="presentation"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                          />
                        </svg>
                      </span>
                    </div>
                  </div>
                </div>
                <div
                  class="ResizingControl-resizeHandle-127 TableHeaderCell-resizeHandle-95"
                >
                  <div
                    class="TableHeaderCell-resizeHandleLine-96 ResizingControl-resizeHandleLine-128 ResizingControl-resizeHandleFirstLine-129"
                  />
                  <div
                    class="TableHeaderCell-resizeHandleLine-96 ResizingControl-resizeHandleLine-128 ResizingControl-resizeHandleSecondLine-130"
                  />
                </div>
              </th>
              <th
                class="MuiTableCell-root-71 MuiTableCell-head-72 TableHeaderCell-cell-94 TableHeaderCell-cellNoUserSelect-98 TableHeaderCell-cellDraggable-99 TableHeaderCell-cellNoWrap-104"
                colspan="1"
                scope="col"
              >
                <div
                  class="TableHeaderCell-container-103"
                >
                  <div
                    class="Content-content-105"
                  >
                    <div
                      class="SortLabel-root-108"
                    >
                      <span
                        class="MuiButtonBase-root-32 MuiTableSortLabel-root-121 SortLabel-sortLabelRoot-110"
                        role="button"
                        tabindex="0"
                        title="Sort"
                      >
                        <span
                          class="Title-title-126"
                        >
                          Reference
                        </span>
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root-35 MuiTableSortLabel-icon-123 MuiTableSortLabel-iconDirectionDesc-124"
                          focusable="false"
                          role="presentation"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                          />
                        </svg>
                      </span>
                    </div>
                  </div>
                </div>
                <div
                  class="ResizingControl-resizeHandle-127 TableHeaderCell-resizeHandle-95"
                >
                  <div
                    class="TableHeaderCell-resizeHandleLine-96 ResizingControl-resizeHandleLine-128 ResizingControl-resizeHandleFirstLine-129"
                  />
                  <div
                    class="TableHeaderCell-resizeHandleLine-96 ResizingControl-resizeHandleLine-128 ResizingControl-resizeHandleSecondLine-130"
                  />
                </div>
              </th>
              <th
                class="MuiTableCell-root-71 MuiTableCell-head-72 TableHeaderCell-cell-94 TableHeaderCell-cellNoUserSelect-98 TableHeaderCell-cellDraggable-99 TableHeaderCell-cellNoWrap-104"
                colspan="1"
                scope="col"
              >
                <div
                  class="TableHeaderCell-container-103"
                >
                  <div
                    class="Content-content-105"
                  >
                    <div
                      class="SortLabel-root-108"
                    >
                      <span
                        class="MuiButtonBase-root-32 MuiTableSortLabel-root-121 SortLabel-sortLabelRoot-110"
                        role="button"
                        tabindex="0"
                        title="Sort"
                      >
                        <span
                          class="Title-title-126"
                        >
                          Length (s)
                        </span>
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root-35 MuiTableSortLabel-icon-123 MuiTableSortLabel-iconDirectionDesc-124"
                          focusable="false"
                          role="presentation"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                          />
                        </svg>
                      </span>
                    </div>
                  </div>
                </div>
                <div
                  class="ResizingControl-resizeHandle-127 TableHeaderCell-resizeHandle-95"
                >
                  <div
                    class="TableHeaderCell-resizeHandleLine-96 ResizingControl-resizeHandleLine-128 ResizingControl-resizeHandleFirstLine-129"
                  />
                  <div
                    class="TableHeaderCell-resizeHandleLine-96 ResizingControl-resizeHandleLine-128 ResizingControl-resizeHandleSecondLine-130"
                  />
                </div>
              </th>
              <th
                class="MuiTableCell-root-71 MuiTableCell-head-72 TableHeaderCell-cell-94 TableHeaderCell-cellNoUserSelect-98 TableHeaderCell-cellDraggable-99 TableHeaderCell-cellNoWrap-104"
                colspan="1"
                scope="col"
              >
                <div
                  class="TableHeaderCell-container-103"
                >
                  <div
                    class="Content-content-105"
                  >
                    <div
                      class="SortLabel-root-108"
                    >
                      <span
                        class="MuiButtonBase-root-32 MuiTableSortLabel-root-121 SortLabel-sortLabelRoot-110"
                        role="button"
                        tabindex="0"
                        title="Sort"
                      >
                        <span
                          class="Title-title-126"
                        >
                          Size (KB)
                        </span>
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root-35 MuiTableSortLabel-icon-123 MuiTableSortLabel-iconDirectionDesc-124"
                          focusable="false"
                          role="presentation"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                          />
                        </svg>
                      </span>
                    </div>
                  </div>
                </div>
                <div
                  class="ResizingControl-resizeHandle-127 TableHeaderCell-resizeHandle-95"
                >
                  <div
                    class="TableHeaderCell-resizeHandleLine-96 ResizingControl-resizeHandleLine-128 ResizingControl-resizeHandleFirstLine-129"
                  />
                  <div
                    class="TableHeaderCell-resizeHandleLine-96 ResizingControl-resizeHandleLine-128 ResizingControl-resizeHandleSecondLine-130"
                  />
                </div>
              </th>
              <th
                class="MuiTableCell-root-71 MuiTableCell-head-72 TableHeaderCell-cell-94 TableHeaderCell-cellNoUserSelect-98 TableHeaderCell-cellDraggable-99 TableHeaderCell-cellNoWrap-104"
                colspan="1"
                scope="col"
              >
                <div
                  class="TableHeaderCell-container-103"
                >
                  <div
                    class="Content-content-105"
                  >
                    <div
                      class="SortLabel-root-108"
                    >
                      <span
                        class="MuiButtonBase-root-32 MuiTableSortLabel-root-121 SortLabel-sortLabelRoot-110"
                        role="button"
                        tabindex="0"
                        title="Sort"
                      >
                        <span
                          class="Title-title-126"
                        >
                          Version
                        </span>
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root-35 MuiTableSortLabel-icon-123 MuiTableSortLabel-iconDirectionDesc-124"
                          focusable="false"
                          role="presentation"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                          />
                        </svg>
                      </span>
                    </div>
                  </div>
                </div>
                <div
                  class="ResizingControl-resizeHandle-127 TableHeaderCell-resizeHandle-95"
                >
                  <div
                    class="TableHeaderCell-resizeHandleLine-96 ResizingControl-resizeHandleLine-128 ResizingControl-resizeHandleFirstLine-129"
                  />
                  <div
                    class="TableHeaderCell-resizeHandleLine-96 ResizingControl-resizeHandleLine-128 ResizingControl-resizeHandleSecondLine-130"
                  />
                </div>
              </th>
              <th
                class="MuiTableCell-root-71 MuiTableCell-head-72 TableHeaderCell-cell-94 TableHeaderCell-cellNoUserSelect-98 TableHeaderCell-cellDraggable-99 TableHeaderCell-cellNoWrap-104"
                colspan="1"
                scope="col"
              >
                <div
                  class="TableHeaderCell-container-103"
                >
                  <div
                    class="Content-content-105"
                  >
                    <div
                      class="SortLabel-root-108"
                    >
                      <span
                        class="MuiButtonBase-root-32 MuiTableSortLabel-root-121 SortLabel-sortLabelRoot-110"
                        role="button"
                        tabindex="0"
                        title="Sort"
                      >
                        <span
                          class="Title-title-126"
                        >
                          Date
                        </span>
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root-35 MuiTableSortLabel-icon-123 MuiTableSortLabel-iconDirectionDesc-124"
                          focusable="false"
                          role="presentation"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                          />
                        </svg>
                      </span>
                    </div>
                  </div>
                </div>
                <div
                  class="ResizingControl-resizeHandle-127 TableHeaderCell-resizeHandle-95"
                >
                  <div
                    class="TableHeaderCell-resizeHandleLine-96 ResizingControl-resizeHandleLine-128 ResizingControl-resizeHandleFirstLine-129"
                  />
                  <div
                    class="TableHeaderCell-resizeHandleLine-96 ResizingControl-resizeHandleLine-128 ResizingControl-resizeHandleSecondLine-130"
                  />
                </div>
              </th>
              <th
                class="MuiTableCell-root-71 MuiTableCell-head-72 TableStubCell-cell-133"
                colspan="1"
                scope="col"
              />
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root-135"
          >
            <tr
              class="MuiTableRow-root-62"
            >
              <td
                class="MuiTableCell-root-71 MuiTableCell-body-73 TableNoDataCell-cell-136"
                colspan="9"
              >
                <div
                  class="TableNoDataCell-textContainer-137"
                >
                  <big
                    class="TableNoDataCell-text-138"
                  >
                    No data
                  </big>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
  <div />
</div>
`;
